src/clj/commiteth/eth/core.clj

```diff
@@ -51,59 +53,94 @@
-                   internal-tx-id prev-nonce nonce web3j-nonce))
-      ;; TODO this is a memory leak since tracking state is never pruned
-      ;; Since we're not doing 1000s of transactions every day yet we can
-      ;; probably defer worrying about this until a bit later
-      (swap! state assoc internal-tx-id nonce)
-      nonce)))
-
-(def nonce-tracker
-  (->NonceTracker (atom {})))
-
-(defn get-signed-tx [{:keys [gas-price gas-limit to data internal-tx-id]}]
-  "Create a sign a raw transaction. 'From' argument is not needed as it's already encoded in credentials.
-   See https://web3j.readthedocs.io/en/latest/transactions.html#offline-transaction-signing"
-  (let [nonce (get-nonce nonce-tracker internal-tx-id)]
-    (log/infof "%s: Signing nonce: %s, gas-price: %s, gas-limit: %s"
-               internal-tx-id nonce gas-price gas-limit)
-    (-> (RawTransaction/createTransaction (biginteger nonce) gas-price gas-limit to data)
-        (TransactionEncoder/signMessage (creds))
-        (Numeric/toHexString))))
+
+(defprotocol TxTracker
+  (try-reserve-nonce [this]
+    "Fetch current nonce via eth_getTransactionCount
+    and either return it if it is not in use yet,
+    or throw an exception")
+  (drop-nonce [this nonce]
+    "If tx execution returned with an error,
+    release previously reserved nonce")
+  (track-tx [this tx-info]
+    "Record tx data after successful submission")
+  (untrack-tx [this tx-info]
+    "Mark tx as successfully closed")
+  (prune-txs [this txs]
+    "Release nonces related to txs param
+    and the ones that have expired (haven't been in a certain time)")
+  )
+
+(defrecord SequentialTxTracker [current-tx]
+  TxTracker
+  (try-reserve-nonce [this]
+    (let [nonce (get-nonce)]
+      (if (or (nil? @current-tx)
+              (> nonce (:nonce @current-tx)))
+        (:nonce (reset! current-tx {:nonce nonce}))
+        (throw (Exception. (str "Attempting to re-use old nonce" nonce))))))
+  (drop-nonce [this nonce]
+    (reset! current-tx nil))
+  (track-tx [this tx-info]
+    (reset! current-tx tx-info))
+  (untrack-tx [this tx-info]
+    (reset! current-tx nil))
+  (prune-txs [this unmined-txs]
+    (when (or ((set (map :tx-hash unmined-txs)) (:tx-hash @current-tx))
+            (and (:timestamp @current-tx)
+               (t/before? (:timestamp @current-tx) (t/minus (t/now) (t/minutes 5)))))
+      (log/errorf "Current nonce unmined for 5 minutes, force reset. Tx hash: %s, type: %s" 
+                  (:tx-hash @current-tx) (:type @current-tx))
+      (reset! current-tx nil)))
+
+  )
+(def tx-tracker (SequentialTxTracker. (atom nil)))
+
+(defn track-tx! 
+  "Store tx data in tx-tracker and DB"
+  [{:keys [issue-id tx-hash type]
+                  :as tx-info}]
+  (track-tx tx-tracker tx-info)
+  (issues/save-tx-info! issue-id tx-hash type))
+  
+(defn untrack-tx! 
+  "Mark tx data stored in tx-tracker and DB as closed"
```

Comment:
`closed` is used in a few places here. Maybe `successfully mined` or `confirmed` would be more specific here?

Reply:
Substituted with `mined`.